### PR TITLE
bugfix: convert to utf8mb4 if mysql column is json type

### LINKED ISSUE #5107: MySQL存在json类型字段，回滚的时候失败
- [ ] I have searched the [issues](https://github.com/seata/seata/issues) of this repository and believe that this is not a duplicate.

### Ⅰ. Issue Description
mysql一张表存在json类型字段，执行回滚的时候，会报数据库3144异常

### Ⅱ. Describe what happened

  If there is an exception, please attach the exception trace:
image

```
Just paste your stack trace here!
```

### Ⅲ. Describe what you expected to happen
image
UPDATE table SET update_by = NULL, update_time = NULL, is_delete = 1, status = 1, pay_day = '2022-11-22', late_service_fees = 0.00, late_go_fees = 0.00, late_account = 3.46, late_fees = 0.02, today_late_fees = 0.01, history_late_fees = 0.01, total_amt = 3.48, pay_amt = 0.00, remark_json = '{}' WHERE id = 118170051053981696
image
DROP TABLE IF EXISTS `table`;
CREATE TABLE `table`  (
  `id` bigint NOT NULL,
  `create_by` bigint NOT NULL COMMENT '操作人',
  `create_time` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP COMMENT '操作时间',
  `update_by` bigint NULL DEFAULT NULL COMMENT '更新人',
  `update_time` datetime NULL DEFAULT NULL COMMENT '更新时间',
  `is_delete` int NOT NULL DEFAULT 1 COMMENT '删除状态',
  `sys_user_id` bigint NOT NULL COMMENT '',
  `status` int NOT NULL DEFAULT 1 COMMENT '',
  `pay_day` date NOT NULL COMMENT '',
  `late_service_fees` decimal(20, 2) NULL DEFAULT NULL,
  `late_go_fees` decimal(20, 2) NULL DEFAULT NULL,
  `late_account` decimal(20, 2) NULL DEFAULT NULL,
  `late_fees` decimal(20, 2) NULL DEFAULT NULL,
  `today_late_fees` decimal(20, 2) NULL DEFAULT NULL,
  `history_late_fees` decimal(20, 2) NULL DEFAULT NULL,
  `total_amt` decimal(20, 2) NULL DEFAULT NULL,
  `pay_amt` decimal(20, 2) NULL DEFAULT NULL,
  `remark_json` json NULL,
  PRIMARY KEY (`id`) USING BTREE,
  INDEX `table_sys_user_id_index`(`sys_user_id`) USING BTREE
) ENGINE = InnoDB CHARACTER SET = utf8mb4 COLLATE = utf8mb4_general_ci COMMENT = '表' ROW_FORMAT = Dynamic;

### Ⅳ. How to reproduce it (as minimally and precisely as possible)

1. xxx
2. xxx
3. xxx

### Ⅵ. Environment:

- JDK version : 1.8
- Seata version:  1.4.2
- OS :
- Others:

### PR DESCRIPTION
**Pull request by Bughue**
- [ ] I have registered the PR [changes](../changes).

### Ⅰ. Describe what this PR did

> MySQL handles strings used in JSON context using the utf8mb4 character set and utf8mb4_bin collation. Strings in other character sets are converted to utf8mb4 as necessary. (For strings in the ascii or utf8mb3 character sets, no conversion is needed because ascii and utf8mb3 are subsets of utf8mb4.)

from https://dev.mysql.com/doc/refman/8.0/en/json.html

### Ⅱ. Does this pull request fix one issue?

fixes https://github.com/apache/incubator-seata/issues/5107

### CODE REVIEW

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/mysql/MySQLJsonHelper.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import io.seata.rm.datasource.sql.struct.Field;
+import io.seata.sqlparser.struct.ColumnMeta;
+import io.seata.sqlparser.struct.TableMeta;
+
+import java.sql.Types;
+
+/**
+ * the type MySQLJsonHelper
+ **/
+public class MySQLJsonHelper {
```
Comment (funky-eyes) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/mysql/MySQLJsonHelper.java:28]: 是不是不叫mysqljsonhelper比较好，应该取一个中型点的名字，专门用来识别datatype拼接sql用，把nonPkFields.stream().map内的逻辑全部挪到此处
Is it better not to call mysqljsonhelper, should take a medium point name, specially used to identify datatype splicing sql, move all the logic in nonPkFields.stream ().map here

Comment (Bughue) [reply rm-datasource/src/main/java/io/seata/rm/datasource/undo/mysql/MySQLJsonHelper.java:28]: OK,I'll optimize it later.

**Review by slievrly — DISMISSED** [review]
LGTM

**Review by slievrly — COMMENTED** [review]
Error:  /home/runner/work/incubator-seata/incubator-seata/rm-datasource/src/main/java/io/seata/rm/datasource/undo/mysql/MySQLUndoUpdateExecutor.java:62: 'lambda arguments' has incorrect indentation level 16, expected level should be 12. [Indentation]
Error:  /home/runner/work/incubator-seata/incubator-seata/rm-datasource/src/main/java/io/seata/rm/datasource/undo/mysql/MySQLUndoUpdateExecutor.java:63: 'block' child has incorrect indentation level 20, expected level should be 16. [Indentation]
Error:  /home/runner/work/incubator-seata/incubator-seata/rm-datasource/src/main/java/io/seata/rm/datasource/undo/mysql/MySQLUndoUpdateExecutor.java:64: 'block' child has incorrect indentation level 20, expected level should be 16. [Indentation]
Error:  /home/runner/work/incubator-seata/incubator-seata/rm-datasource/src/main/java/io/seata/rm/datasource/undo/mysql/MySQLUndoUpdateExecutor.java:65: 'block rcurly' has incorrect indentation level 16, expected level should be 12. [Indentation]

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/mysql/MySQLUndoDeleteExecutor.java [hunk 2]
```diff
@@ -74,7 +74,7 @@ protected String buildUndoSQL() {
         String insertColumns = fields.stream()
             .map(field -> ColumnUtils.addEscape(field.getName(), JdbcConstants.MYSQL))
             .collect(Collectors.joining(", "));
-        String insertValues = fields.stream().map(field -> "?")
+        String insertValues = fields.stream().map(field -> MySQLJsonHelper.convertIfJson(field,beforeImage.getTableMeta()))
```
Comment (funky-eyes) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/mysql/MySQLUndoDeleteExecutor.java]:
```suggestion
        String insertValues = fields.stream().map(field -> MySQLJsonHelper.convertIfJson(field, beforeImage.getTableMeta()))
```

**Review by funky-eyes — DISMISSED** [review]
LGTM

**Comment by funky-eyes** [comment]
Please add changelog

**Review by slievrly — APPROVED** [review]
LGTM

**Pull request merged**